Installation issue: octave

### Steps to reproduce the issue

```console
$ spack spec -I <spec>
Input spec
--------------------------------
octave

Concretized
--------------------------------
fctts7i  octave@8.2.0%gcc@13.1.0~arpack+bz2~curl~fftw~fltk~fontconfig~freetype~gl2ps~glpk~gnuplot~hdf5~jdk~llvm~magick~opengl~qhull~qrupdate~qscintilla~qt+readline~suitesparse~zlib build_system=autotools arch=linux-rhel8-icelake
6ox4r3q      ^bzip2@1.0.8%gcc@13.1.0~debug~pic+shared build_system=generic arch=linux-rhel8-icelake
3wy6t5d          ^diffutils@3.9%gcc@13.1.0 build_system=autotools arch=linux-rhel8-icelake
wbapu7t              ^libiconv@1.17%gcc@13.1.0 build_system=autotools libs=shared,static arch=linux-rhel8-icelake
yve3gnx      ^openblas@0.3.23%gcc@13.1.0~bignuma~consistent_fpcsr+fortran+ilp64+locking+pic+shared build_system=makefile symbol_suffix=64_ threads=none arch=linux-rhel8-icelake
dzy6eq5          ^perl@5.36.0%gcc@13.1.0+cpanm+open+shared+threads build_system=generic arch=linux-rhel8-icelake
pps42zb              ^berkeley-db@18.1.40%gcc@13.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rhel8-icelake
aqrpxbj              ^gdbm@1.23%gcc@13.1.0 build_system=autotools arch=linux-rhel8-icelake
oelwntp              ^zlib@1.2.13%gcc@13.1.0+optimize+pic+shared build_system=makefile arch=linux-rhel8-icelake
mveoqwe      ^pcre@8.45%gcc@13.1.0~jit+multibyte+utf build_system=autotools arch=linux-rhel8-icelake
ju7fvw5      ^pkgconf@1.9.5%gcc@13.1.0 build_system=autotools arch=linux-rhel8-icelake
g23axr6      ^readline@8.2%gcc@13.1.0 build_system=autotools patches=bbf97f1 arch=linux-rhel8-icelake
u66fwxi          ^ncurses@6.4%gcc@13.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-rhel8-icelake
y6b3j64      ^texinfo@7.0.3%gcc@13.1.0 build_system=autotools arch=linux-rhel8-icelake
hqbacim          ^gettext@0.21.1%gcc@13.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-rhel8-icelake
qnbechw              ^libxml2@2.10.3%gcc@13.1.0~python build_system=autotools arch=linux-rhel8-icelake
f3l3yvj              ^tar@1.34%gcc@13.1.0 build_system=autotools zip=pigz arch=linux-rhel8-icelake
czvo6b7                  ^pigz@2.7%gcc@13.1.0 build_system=makefile arch=linux-rhel8-icelake
mxsgv4d                  ^zstd@1.5.5%gcc@13.1.0+programs build_system=makefile compression=none libs=shared,static arch=linux-rhel8-icelake
4jfeqd7              ^xz@5.4.1%gcc@13.1.0~pic build_system=autotools libs=shared,static arch=linux-rhel8-icelake

```


### Error message

<details><summary>Error message</summary>

<pre>

==> Installing octave-8.2.0-6t645kj6pqfzyhzdw6tlyrnnx26qyltc
==> No binary for octave-8.2.0-6t645kj6pqfzyhzdw6tlyrnnx26qyltc found: installing from source
==> Using cached archive: /home/sw/spack-v0.20.1-a/var/spack/cache/_source-cache/archive/57/57d17f918a940d38ca3348211e110b34d735a322a87db71c177c4692a49a9c84.tar.gz
==> Ran patch() for octave
==> octave: Executing phase: 'autoreconf'
==> octave: Executing phase: 'configure'
==> octave: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

23 errors found in build log:
     7009     In file included from ./libinterp/octave-value/ov-fcn.h:34:
     7010     In file included from ./liboctave/util/str-vec.h:35:
     7011     In file included from ./liboctave/array/Array.h:40:
     7012     In file included from ./liboctave/array/idx-vector.h:40:
     7013     In file included from ./liboctave/util/oct-inttypes.h:37:
     7014     liboctave/numeric/lo-mappers.h:181:39: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]
  >> 7015     liboctave/numeric/lo-mappers.h:203:39: inline bool isnan (double x) { return std::isnan (x); }warning:
     7016     explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]                                      ^~~~~~~~~~~~~~
     7017
     7018     inline bool isinf (double x) { return std::isinf (x); }
     7019                                           ^~~~~~~~~~~~~~
     7020     liboctave/numeric/lo-mappers.h:182:38: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]liboctave/numeric/lo-mappers.h
     7021     :204:38: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]

     ...

     40079    ./liboctave/util/lo-ieee.h:84:53: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]
     40080    inline int __lo_ieee_float_isnan (float x) { return std::isnan (x); }
     40081                                                        ^~~~~~~~~~~~~~
     40082    ./liboctave/util/lo-ieee.h:86:53: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]
     40083    inline int __lo_ieee_float_isinf (float x) { return std::isinf (x); }
     40084                                                        ^~~~~~~~~~~~~~
  >> 40085    liboctave/util/lo-regexp.cc:74:9: error: unknown type name 'pcre'
     40086    typedef pcre octave_pcre_code;
     40087            ^
     40088    In file included from liboctave/util/str-vec.cc:41:
     40089    In file included from ./liboctave/util/cmd-edit.h:36:
     40090    In file included from ./liboctave/util/str-vec.h:35:
     40091    In file included from ./liboctave/array/Array.h:40:

     ...
</pre></details>


### Information on your system

* **Spack:** 0.20.1 (e8658d6493887ef702dd38f0e9ee5870a1651c1e)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-icelake
* **Concretizer:** clingo

### Additional information

Trying to debug the issue with a build environment as suggested [here](https://app.slack.com/client/T5URKQG73/C01UZDL35C4/thread/C01UZDL35C4-1690324265.729769) (if slack links work, sigh) by @becker33 seems to solve the problem:

```
[root@co49svhead01 spack-v0.20.1-a]# spack build-env octave -- bash
[root@co49svhead01 spack-v0.20.1-a]# cd /tmp/root/spack-stage/spack-stage-octave-8.2.0-6t645kj6pqfzyhzdw6tlyrnnx26qyltc/spack-src/
[root@co49svhead01 spack-src]# /bin/sh ./libtool  --tag=CXX   --mode=compile /home/sw/spack-v0.20.1-a/lib/spack/env/oneapi/icpx -DHAVE_CONFIG_H -I.  -Iliboctave -I./liboctave -I./liboctave/array -Iliboctave/numeric -I./liboctave/numeric -Iliboctave/operators -I./liboctave/operators -I./liboctave/system -I./liboctave/util -I./libinterp/octave-value -Ilibinterp -I./libinterp -I./libinterp/operators -Ilibinterp/parse-tree -I./libinterp/parse-tree -Ilibinterp/corefcn -I./libinterp/corefcn -I./liboctave/wrappers -I/home/sw/spack-v0.20.1-a/opt/spack/linux-rhel8-icelake/oneapi-2023.2.0/hdf5-1.14.1-2-vcoeh2dgmcuhdyi3asgfhgh3abln3yje/include     -fPIC -pthread -fopenmp -Wall -W -Wshadow -Woverloaded-virtual -Wold-style-cast -Wformat -Wpointer-arith -Wwrite-strings -Wcast-align -Wcast-qual  -O2 -MT libinterp/dldfcn/__fltk_uigetfile___la-__fltk_uigetfile__.lo -MD -MP -MF libinterp/dldfcn/.deps/__fltk_uigetfile___la-__fltk_uigetfile__.Tpo -c -o libinterp/dldfcn/__fltk_uigetfile___la-__fltk_uigetfile__.lo `test -f 'libinterp/dldfcn/__fltk_uigetfile__.cc' || echo './'`libinterp/dldfcn/__fltk_uigetfile__.cc
libtool: compile:  /home/sw/spack-v0.20.1-a/lib/spack/env/oneapi/icpx -DHAVE_CONFIG_H -I. -Iliboctave -I./liboctave -I./liboctave/array -Iliboctave/numeric -I./liboctave/numeric -Iliboctave/operators -I./liboctave/operators -I./liboctave/system -I./liboctave/util -I./libinterp/octave-value -Ilibinterp -I./libinterp -I./libinterp/operators -Ilibinterp/parse-tree -I./libinterp/parse-tree -Ilibinterp/corefcn -I./libinterp/corefcn -I./liboctave/wrappers -I/home/sw/spack-v0.20.1-a/opt/spack/linux-rhel8-icelake/oneapi-2023.2.0/hdf5-1.14.1-2-vcoeh2dgmcuhdyi3asgfhgh3abln3yje/include -fPIC -pthread -fopenmp -Wall -W -Wshadow -Woverloaded-virtual -Wold-style-cast -Wformat -Wpointer-arith -Wwrite-strings -Wcast-align -Wcast-qual -O2 -MT libinterp/dldfcn/__fltk_uigetfile___la-__fltk_uigetfile__.lo -MD -MP -MF libinterp/dldfcn/.deps/__fltk_uigetfile___la-__fltk_uigetfile__.Tpo -c libinterp/dldfcn/__fltk_uigetfile__.cc  -fPIC -DPIC -o libinterp/dldfcn/.libs/__fltk_uigetfile___la-__fltk_uigetfile__.o
In file included from libinterp/dldfcn/__fltk_uigetfile__.cc:49:
In file included from ./liboctave/array/dMatrix.h:31:
In file included from liboctave/numeric/DET.h:33:
In file included from liboctave/numeric/lo-mappers.h:35:
./liboctave/util/lo-ieee.h:72:48: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]
inline int __lo_ieee_isnan (double x) { return std::isnan (x); }
                                               ^~~~~~~~~~~~~~
./liboctave/util/lo-ieee.h:74:48: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]
inline int __lo_ieee_isinf (double x) { return std::isinf (x); }
                                               ^~~~~~~~~~~~~~
./liboctave/util/lo-ieee.h:84:53: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]
inline int __lo_ieee_float_isnan (float x) { return std::isnan (x); }
                                                    ^~~~~~~~~~~~~~
./liboctave/util/lo-ieee.h:86:53: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]
inline int __lo_ieee_float_isinf (float x) { return std::isinf (x); }
                                                    ^~~~~~~~~~~~~~
In file included from libinterp/dldfcn/__fltk_uigetfile__.cc:49:
In file included from ./liboctave/array/dMatrix.h:31:
In file included from liboctave/numeric/DET.h:33:
liboctave/numeric/lo-mappers.h:181:39: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]
inline bool isnan (double x) { return std::isnan (x); }
                                      ^~~~~~~~~~~~~~
liboctave/numeric/lo-mappers.h:182:38: warning: explicit comparison with NaN in fast floating point mode [-Wtautological-constant-compare]
inline bool isnan (float x) { return std::isnan (x); }
                                     ^~~~~~~~~~~~~~
liboctave/numeric/lo-mappers.h:203:39: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]
inline bool isinf (double x) { return std::isinf (x); }
                                      ^~~~~~~~~~~~~~
liboctave/numeric/lo-mappers.h:204:38: warning: explicit comparison with infinity in fast floating point mode [-Wtautological-constant-compare]
inline bool isinf (float x) { return std::isinf (x); }
                                     ^~~~~~~~~~~~~~
8 warnings generated.
```

So hopefully this gives some clues on what the problem is which I can't figure out
@mtmiller  @siko1056

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/octave/package.py
+++ b/var/spack/repos/builtin/packages/octave/package.py
@@ -351,9 +351,8 @@ class Octave(AutotoolsPackage, GNUMirrorPackage):
         else:
             config_args.append("--without-z")
 
-        if "~pcre2" in spec:
+        if spec.satisfies("~pcre2"):
             config_args.append("--without-pcre2")
-
         # If 64-bit BLAS is used:
         if (
             spec.satisfies("^openblas+ilp64")

--- a/var/spack/repos/builtin/packages/octave/package.py
+++ b/var/spack/repos/builtin/packages/octave/package.py
@@ -82,7 +82,8 @@ class Octave(AutotoolsPackage, GNUMirrorPackage):
     depends_on("lapack")
     # Octave does not configure with sed from darwin:
     depends_on("sed", when=sys.platform == "darwin", type="build")
-    depends_on("pcre")
+    depends_on("pcre", when="@:7")
+    depends_on("pcre2", when="@8:")
     depends_on("pkgconfig", type="build")
     depends_on("texinfo", type="build")
 

--- a/var/spack/repos/builtin/packages/octave/package.py
+++ b/var/spack/repos/builtin/packages/octave/package.py
@@ -70,6 +70,7 @@ class Octave(AutotoolsPackage, GNUMirrorPackage):
     variant("jdk", default=False, description="Use JDK")
     variant("llvm", default=False, description="Use LLVM")
     variant("opengl", default=False, description="Use OpenGL")
+    variant("pcre2", default=True, when="@8:", description="Use PCRE2 instead of PCRE")
     variant("qhull", default=False, description="Use qhull")
     variant("qrupdate", default=False, description="Use qrupdate")
     variant("qscintilla", default=False, description="Use QScintill")
@@ -83,7 +84,8 @@ class Octave(AutotoolsPackage, GNUMirrorPackage):
     # Octave does not configure with sed from darwin:
     depends_on("sed", when=sys.platform == "darwin", type="build")
     depends_on("pcre", when="@:7")
-    depends_on("pcre2", when="@8:")
+    depends_on("pcre", when="~pcre2")
+    depends_on("pcre2", when="+pcre2")
     depends_on("pkgconfig", type="build")
     depends_on("texinfo", type="build")
 
@@ -348,6 +350,9 @@ class Octave(AutotoolsPackage, GNUMirrorPackage):
             )
         else:
             config_args.append("--without-z")
+
+        if "~pcre2" in spec:
+            config_args.append("--without-pcre2")
 
         # If 64-bit BLAS is used:
         if (

